ccl/changefeedccl: TestAlterChangefeedAddTargetsDuringBackfill failed

ccl/changefeedccl.TestAlterChangefeedAddTargetsDuringBackfill [failed](https://tanzanite.cluster.engflow.com/invocations/default/ed4c1d3d-62fb-4e38-9fa4-6341741d7fc0?testReportRun=3&testReportShard=8&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [f0116ea373a2b87155e7f0264df4f783ce177360](https://github.com/cockroachdb/cockroach/commits/f0116ea373a2b87155e7f0264df4f783ce177360):


```
=== RUN   TestAlterChangefeedAddTargetsDuringBackfill
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestAlterChangefeedAddTargetsDuringBackfill1577098480
    test_log_scope.go:81: use -show-logs to present logs inline
    alter_changefeed_test.go:1449: -- test log scope end --
test logs left over in: outputs.zip/logTestAlterChangefeedAddTargetsDuringBackfill1577098480
--- FAIL: TestAlterChangefeedAddTargetsDuringBackfill (184.01s)
=== RUN   TestAlterChangefeedAddTargetsDuringBackfill/enterprise
    helpers_test.go:867: making server as system tenant
    helpers_test.go:943: making enterprise feed factory
    helpers_test.go:957: pgURL enterprise root
    helpers_test.go:682: waiting for checkpoint
    helpers_test.go:682: waiting for checkpoint
    helpers_test.go:682: waiting for checkpoint
    helpers_test.go:685: read checkpoint spans:<key:"\362\211\366\232\210" end_key:"\362\211\366\333\210" > timestamp:<wall_time:1710925058292374311 > 
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:704: could not read highwater
    --- FAIL: TestAlterChangefeedAddTargetsDuringBackfill/enterprise (184.00s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=8</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAlterChangefeedAddTargetsDuringBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-36850

ccl/changefeedccl.TestAlterChangefeedAddTargetsDuringBackfill [failed](https://tanzanite.cluster.engflow.com/invocations/default/613e74e8-563c-43e0-b135-d85141636ac3?testReportRun=1&testReportShard=8&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [3060756a084cc437114d64265e79ace5720a7317](https://github.com/cockroachdb/cockroach/commits/3060756a084cc437114d64265e79ace5720a7317):


```
=== RUN   TestAlterChangefeedAddTargetsDuringBackfill
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestAlterChangefeedAddTargetsDuringBackfill2874242043
    test_log_scope.go:81: use -show-logs to present logs inline
    alter_changefeed_test.go:1449: -- test log scope end --
test logs left over in: outputs.zip/logTestAlterChangefeedAddTargetsDuringBackfill2874242043
--- FAIL: TestAlterChangefeedAddTargetsDuringBackfill (106.60s)
=== RUN   TestAlterChangefeedAddTargetsDuringBackfill/pulsar
    helpers_test.go:867: making server as system tenant
    helpers_test.go:943: making pulsar feed factory
    helpers_test.go:682: waiting for checkpoint
    helpers_test.go:682: waiting for checkpoint
    helpers_test.go:682: waiting for checkpoint
    helpers_test.go:685: read checkpoint spans:<key:"\362\211\303\210" end_key:"\362\211\366\202\210" > spans:<key:"\362\211\366\274\210" end_key:"\362\211\367\0018\210" > timestamp:<wall_time:1711012232412208360 > 
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:697: waiting for highwater
    helpers_test.go:704: could not read highwater
    --- FAIL: TestAlterChangefeedAddTargetsDuringBackfill/pulsar (106.60s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=8</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAlterChangefeedAddTargetsDuringBackfill.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/ccl/changefeedccl/alter_changefeed_test.go
+++ b/pkg/ccl/changefeedccl/alter_changefeed_test.go
@@ -12,6 +12,7 @@ import (
 	"context"
 	gosql "database/sql"
 	"fmt"
+	"math/rand"
 	"net/url"
 	"sync/atomic"
 	"testing"
@@ -1318,8 +1319,11 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	rnd, _ := randutil.NewTestRand()
-	var rndMu syncutil.Mutex
+	var rndMu struct {
+		syncutil.Mutex
+		rnd *rand.Rand
+	}
+	rndMu.rnd, _ = randutil.NewTestRand()
 	const maxCheckpointSize = 1 << 20
 	const numRowsPerTable = 1000
 
@@ -1340,11 +1344,19 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 			Changefeed.(*TestingKnobs)
 
 		// Ensure Scan Requests are always small enough that we receive multiple
-		// resolvedFoo events during a backfill
+		// resolvedFoo events during a backfill.
+		const maxBatchSize = numRowsPerTable / 5
 		knobs.FeedKnobs.BeforeScanRequest = func(b *kv.Batch) error {
 			rndMu.Lock()
 			defer rndMu.Unlock()
-			b.Header.MaxSpanRequestKeys = 1 + rnd.Int63n(100)
+			// We don't want batch sizes that are too small because they could cause
+			// the initial scan to take too long, leading to the waitForHighwater
+			// call below to time out. The formula below is completely arbitrary and
+			// was chosen to ensure that the batch sizes aren't too small (i.e. in
+			// the 1-2 digit range) but are still small enough that at least a few
+			// batches will be necessary.
+			b.Header.MaxSpanRequestKeys = maxBatchSize/2 + rndMu.rnd.Int63n(maxBatchSize/2)
+			t.Logf("set max span request keys: %d", b.Header.MaxSpanRequestKeys)
 			return nil
 		}
 
@@ -1365,7 +1377,7 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 				return false, nil
 			}
 			if haveGaps {
-				return rnd.Intn(10) > 7, nil
+				return rndMu.rnd.Intn(10) > 7, nil
 			}
 			haveGaps = true
 			return true, nil
@@ -1415,13 +1427,12 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 		var checkpoint roachpb.SpanGroup
 		checkpoint.Add(jobCheckpoint.Spans...)
 
-		waitForJobStatus(sqlDB, t, jobFeed.JobID(), `paused`)
-
 		sqlDB.Exec(t, fmt.Sprintf(`ALTER CHANGEFEED %d ADD bar WITH initial_scan`, jobFeed.JobID()))
 
 		// Collect spans we attempt to resolve after when we resume.
 		var resolvedFoo []roachpb.Span
 		knobs.FilterSpanWithMutation = func(r *jobspb.ResolvedSpan) (bool, error) {
+			t.Logf("resolved span: %#v", r)
 			if !r.Span.Equal(fooTableSpan) {
 				resolvedFoo = append(resolvedFoo, r.Span)
 			}
@@ -1430,6 +1441,7 @@ func TestAlterChangefeedAddTargetsDuringBackfill(t *testing.T) {
 
 		require.NoError(t, jobFeed.Resume())
 
+		// Wait for highwater to be set, which signifies that the initial scan is complete.
 		waitForHighwater(t, jobFeed, registry)
 
 		// At this point, highwater mark should be set, and previous checkpoint should be gone.

